ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound/State.hs

```diff
@@ -224,48 +223,47 @@ instance ( NoThunks peeraddr
 -- Pure public API
 --
 
--- | Check if a peer can acknowledge at least one `txid`.
---
-hasTxIdsToAcknowledge
-    :: forall peeraddr txid tx.
-       Ord txid
-    => SharedTxState peeraddr txid tx
-    -> PeerTxState txid tx
-    -> Bool
-hasTxIdsToAcknowledge
-    SharedTxState { bufferedTxs }
-    PeerTxState { unacknowledgedTxIds, unknownTxs }
-    =
-    -- We just need to look at the front of the unacknowledged `txid`s.
-    case unacknowledgedTxIds of
-      txid StrictSeq.:<| _ -> txid `Map.member` bufferedTxs
-                           || txid `Set.member` unknownTxs
-      _                    -> False
-
-
 acknowledgeTxIds
     :: forall peeraddr tx txid.
        Ord txid
-    => SharedTxState peeraddr txid tx
+    => TxDecisionPolicy
+    -> SharedTxState peeraddr txid tx
     -> PeerTxState txid tx
-    -> (NumTxIdsToAck, [tx], RefCountDiff txid, PeerTxState txid tx)
+    -> (NumTxIdsToAck, NumTxIdsToReq, [tx], RefCountDiff txid, PeerTxState txid tx)
     -- ^ number of txid to acknowledge, txids to acknowledge with multiplicities,
     -- updated PeerTxState.
 {-# INLINE acknowledgeTxIds #-}
 
 acknowledgeTxIds
+    TxDecisionPolicy { maxNumTxIdsToRequest,
+                       maxUnacknowledgedTxIds }
     SharedTxState { bufferedTxs }
     ps@PeerTxState { availableTxIds,
                      unacknowledgedTxIds,
-                     unknownTxs }
+                     unknownTxs,
+                     requestedTxIdsInflight }
     =
-    ( fromIntegral $ StrictSeq.length acknowledgedTxIds
-    , txsToMempool
-    , refCountDiff
-    , ps { unacknowledgedTxIds = unacknowledgedTxIds',
-           availableTxIds      = availableTxIds',
-           unknownTxs          = unknownTxs' }
-    )
+    -- We can only acknowledge txids when we can request new ones, since
+    -- a `MsgRequestTxIds` for 0 txids is a protocol error.
+    if txIdsToRequest > 0
+      then
+      ( txIdsToAcknowledge
+      , txIdsToRequest
+      , txsToMempool
+      , refCountDiff
+      , ps { unacknowledgedTxIds    = unacknowledgedTxIds',
+             availableTxIds         = availableTxIds',
+             unknownTxs             = unknownTxs',
+             requestedTxIdsInflight = requestedTxIdsInflight
+                                    + txIdsToRequest }
+      )
+      else
+      ( 0
+      , 0
+      , []
```

Comment:
@bolt12 I think this is too conservative, why don't we return `txsToMempool` here?